File: 哆啦A梦跑酷/Doraemon Parkour-游戏源码/DoraemonParkour/Models/GameAppState.swift
import Foundation
import UIKit

enum ChestOverlay: Equatable {
    case reward(ChestKind, RewardBundle)
    case missingKey(ChestKind)
    case missingChest(ChestKind)
}

enum ShopDetail: Equatable {
    case character(String)
    case item(String)
    case chest(ChestKind)
    case key(KeyKind)
}

@MainActor
final class GameAppState: ObservableObject {
    @Published var screen: AppScreen = .loading
    @Published var profile: PlayerProfile
    @Published var selectedLevel = 1
    @Published var inventoryCategory: InventoryCategory = .characters
    @Published var shopCategory: ShopCategory = .characters
    @Published var selectedCharacterID = "doraemon"
    @Published var selectedItemID = "energy_milk"
    @Published var selectedChest: ChestKind = .wood
    @Published var shopDetail: ShopDetail?
    @Published var chestOverlay: ChestOverlay?
    @Published var showRecharge = false
    @Published var showAbout = false
    @Published var showPrivacy = false
    @Published var showSettingsOverlay = false
    @Published var showInventoryOverlay = false
    @Published var toast: String?
    @Published var latestResult: RunSummary?
    @Published var runSessionID = UUID()

    let audio = AudioManager()
    @Published private(set) var returnScreen: AppScreen = .home

    init() {
        profile = SaveStore.load()
        selectedCharacterID = profile.selectedCharacter
        #if DEBUG
        if ProcessInfo.processInfo.arguments.contains("-uiTestLoading") {
            screen = .loading
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestSuccess") {
            screen = .run
            selectedLevel = 1
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestHome") {
            screen = .home
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestLevels") {
            screen = .levels
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestRunQuickbar") ||
                    ProcessInfo.processInfo.arguments.contains("-uiTestPause") ||
                    ProcessInfo.processInfo.arguments.contains("-uiTestFlight") ||
                    ProcessInfo.processInfo.arguments.contains("-uiTestFlightForward") ||
                    ProcessInfo.processInfo.arguments.contains("-uiTestItemEffect") {
            profile.loadout = ["energy_milk", "shield", "flight_boots"]
            screen = .run
            selectedLevel = 1
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestFailureResult") {
            screen = .result
            latestResult = RunSummary(level: 1, success: false, stars: 0, score: 620, coins: 18, diamonds: 1, itemRewards: ["energy_milk": 1, "mini_dorayaki": 1], chestRewards: [:], keyRewards: [:])
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestSuccessResult") {
            screen = .result
            latestResult = RunSummary(level: 1, success: true, stars: 3, score: 1375, coins: 125, diamonds: 3, itemRewards: ["energy_milk": 1, "shield": 1], chestRewards: [ChestKind.wood.rawValue: 1], keyRewards: [:])
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestInventoryCharacters") {
            screen = .inventory
            inventoryCategory = .characters
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestInventoryItems") {
            screen = .inventory
            inventoryCategory = .items
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestInventoryChests") {
            screen = .inventory
            inventoryCategory = .chests
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestShopCharacters") {
            screen = .shop
            shopCategory = .characters
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestShopItems") {
            screen = .shop
            shopCategory = .items
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestShopChests") {
            screen = .shop
            shopCategory = .chests
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestShopDetail") {
            screen = .shop
            shopDetail = .character("nobita")
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestItemDetail") {
            screen = .shop
            shopCategory = .items
            shopDetail = .item("speed_shoes")
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestChestDetail") {
            screen = .shop
            shopCategory = .chests
            shopDetail = .chest(.wood)
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestKeyDetail") {
            screen = .shop
            shopCategory = .chests
            shopDetail = .key(.copper)
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestSettings") {
            screen = .home
            showSettingsOverlay = true
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestAbout") {
            screen = .home
            showSettingsOverlay = true
            showAbout = true
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestPrivacy") {
            screen = .home
            showSettingsOverlay = true
            showPrivacy = true
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestChestReward") {
            screen = .inventory
            inventoryCategory = .chests
            chestOverlay = .reward(.wood, RewardBundle(coins: 55, items: ["energy_milk": 1]))
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestMissingKey") {
            screen = .inventory
            inventoryCategory = .chests
            chestOverlay = .missingKey(.wood)
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestMissingChest") {
            screen = .inventory
            inventoryCategory = .chests
            chestOverlay = .missingChest(.wood)
        } else if ProcessInfo.processInfo.arguments.contains("-uiTestRecharge") {
            screen = .shop
            showRecharge = true
        }
        #endif
    }

    var selectedCharacter: CharacterDefinition { GameData.character(profile.selectedCharacter) }

    func completeLoading() { navigate(.home) }

    func navigate(_ target: AppScreen) {
        screen = target
        shopDetail = nil
        chestOverlay = nil
        switch target {
        case .run: audio.playBGM("bgm_run", settings: profile.settings)
        case .result: audio.playBGM("bgm_result", settings: profile.settings)
        case .loading: break
        default: audio.playBGM("bgm_home", settings: profile.settings)
        }
    }

    func openSettings(from source: AppScreen) { returnScreen = source; showSettingsOverlay = true }
    func closeSettings() { showSettingsOverlay = false }
    func openInventory(from source: AppScreen) {
        returnScreen = source
        if source == .run { showInventoryOverlay = true } else { navigate(.inventory) }
    }
    func openShop(from source: AppScreen, category: ShopCategory? = nil) { returnScreen = source; if let category { shopCategory = category }; navigate(.shop) }
    func backFromSection() {
        // Returning from a nested section can leave the current section stored
        // as its own destination. Fall back to home instead of navigating to
        // the same screen and making the back button appear unresponsive.
        if returnScreen == screen, screen == .inventory || screen == .shop {
            navigate(.home)
        } else {
            navigate(returnScreen)
        }
    }

    func startLevel(_ level: Int) {
        guard level <= profile.unlockedLevel else { showToast("该关卡尚未解锁") ; return }
        selectedLevel = level
        runSessionID = UUID()
        navigate(.run)
    }

    func restartLevel() { runSessionID = UUID(); navigate(.run) }

    func toggleMusic() {
        profile.settings.musicVolume = profile.settings.musicVolume == 0 ? 8 : 0
        persistSettings()
    }

    func changeVolume(music: Bool, delta: Int) {
        if music { profile.settings.musicVolume = min(10, max(0, profile.settings.musicVolume + delta)) }
        else { profile.settings.soundVolume = min(10, max(0, profile.settings.soundVolume + delta)) }
        persistSettings()
    }

    func resetSettings() { profile.settings = SettingsProfile(); persistSettings(); showToast("已恢复默认设置") }
    func clearCache() { profile.lastRunSummary = nil; SaveStore.save(profile); showToast("缓存已清理，进度已保留") }

    func persistSettings() {
        SaveStore.save(profile)
        audio.update(settings: profile.settings)
    }

    func selectCharacter(_ id: String) {
        let character = GameData.character(id)
        guard profile.unlockedCharacters.contains(id) else { openShop(from: .inventory, category: .characters); shopDetail = .character(id); return }
        guard character.availableInRun else { showToast("动作素材待补，当前角色暂不可出战"); return }
        profile.selectedCharacter = id
        selectedCharacterID = id
        save()
        showToast("已使用\(character.name)")
    }

    func carry(_ itemID: String) {
        guard let item = GameData.item(itemID), (profile.itemInventory[itemID] ?? 0) > 0 else { showToast("背包中没有该道具"); return }
        if profile.loadout.contains(itemID) { return }
        if let owner = item.owner, owner != profile.selectedCharacter { showToast("非当前角色专属"); return }
        guard profile.loadout.count < 3 else { showToast("携带已满，请先取消一格"); return }
        profile.itemInventory[itemID, default: 0] -= 1
        profile.loadout.append(itemID)
        save()
    }

    func cancelCarry(_ itemID: String) {
        guard let index = profile.loadout.firstIndex(of: itemID) else { return }
        profile.loadout.remove(at: index)
        profile.itemInventory[itemID, default: 0] += 1
        save()
    }

    func sell(_ itemID: String) {
        guard let item = GameData.item(itemID), (profile.itemInventory[itemID] ?? 0) > 0 else { showToast("没有可出售的道具"); return }
        profile.itemInventory[itemID, default: 0] -= 1
        profile.coins += item.sellPrice
        save()
        audio.playSFX("sfx_confirm", settings: profile.settings)
        showToast("出售成功，获得 \(item.sellPrice) 金币")
    }

    func purchase(_ detail: ShopDetail) {
        var currency: CurrencyKind = .coins
        var price = 0
        switch detail {
        case .character(let id):
            let value = GameData.character(id)
            guard value.availableInRun else { showToast("动作素材待补，暂未开放购买"); return }
            guard !profile.unlockedCharacters.contains(id) else { showToast("角色已拥有"); return }
            currency = .diamonds; price = value.price
        case .item(let id):
            guard let item = GameData.item(id) else { return }
            price = item.buyPrice
        case .chest(let kind):
            let chest = GameData.chest(kind); currency = chest.currency; price = chest.price
        case .key(let kind):
            guard let key = GameData.keys.first(where: { $0.kind == kind }) else { return }
            guard (profile.keyInventory[kind.rawValue] ?? 0) < 99 else { showToast("该钥匙已达 99 把上限"); return }
            currency = key.currency; price = key.price
        }
        guard spend(price, currency: currency) else {
            GameHaptics.notification(.warning, enabled: profile.settings.vibrationEnabled)
            showToast("\(currency.rawValue)不足")
            return
        }
        switch detail {
        case .character(let id): profile.unlockedCharacters.insert(id)
        case .item(let id): profile.itemInventory[id, default: 0] += 1
        case .chest(let kind): profile.chestInventory[kind.rawValue, default: 0] += 1
        case .key(let kind): profile.keyInventory[kind.rawValue, default: 0] += 1
        }
        save()
        audio.playSFX("sfx_confirm", settings: profile.settings)
        GameHaptics.notification(.success, enabled: profile.settings.vibrationEnabled)
        showToast("购买成功")
    }

    private func spend(_ value: Int, currency: CurrencyKind) -> Bool {
        if currency == .coins { guard profile.coins >= value else { return false }; profile.coins -= value }
        else { guard profile.diamonds >= value else { return false }; profile.diamonds -= value }
        return true
    }

    func openChest(_ kind: ChestKind) {
        let chest = GameData.chest(kind)
        guard (profile.chestInventory[kind.rawValue] ?? 0) > 0 else {
            GameHaptics.notification(.warning, enabled: profile.settings.vibrationEnabled)
            chestOverlay = .missingChest(kind)
            return
        }
        guard (profile.keyInventory[chest.key.rawValue] ?? 0) > 0 else {
            GameHaptics.notification(.warning, enabled: profile.settings.vibrationEnabled)
            chestOverlay = .missingKey(kind)
            return
        }
        profile.chestInventory[kind.rawValue, default: 0] -= 1
        profile.keyInventory[chest.key.rawValue, default: 0] -= 1
        let reward = rollChest(kind)
        apply(reward)
        save()
        audio.playSFX("sfx_chest", settings: profile.settings)
        GameHaptics.notification(.success, enabled: profile.settings.vibrationEnabled)
        chestOverlay = .reward(kind, reward)
    }

    private func rollChest(_ kind: ChestKind) -> RewardBundle {
        var reward = RewardBundle()
        switch kind {
        case .wood:
            reward.coins = Int.random(in: 40...70)
            reward.items[Bool.random() ? "mini_dorayaki" : "energy_milk"] = 1
        case .silver:
            reward.coins = Int.random(in: 80...130)
            reward.items[["classic_dorayaki", "magnet", "shield", "energy_milk"].randomElement()!] = 1
            if Int.random(in: 0..<100) < 15 { reward.diamonds = 1 }
            if Int.random(in: 0..<100) < 3 { reward.character = "nobita" }
        case .gold:
            reward.coins = Int.random(in: 140...220)
            reward.items[["luxury_dorayaki", "flight_boots", "leap_drink", "lucky_drink", "first_aid"].randomElement()!] = 1
            reward.diamonds = [1, 2, 2, 3].randomElement()!
            if Int.random(in: 0..<100) < 6 { reward.character = ["nobita", "shizuka"].randomElement()! }
        case .purple:
            reward.coins = Int.random(in: 180...280)
            reward.items[["flight_boots", "lucky_drink", "first_aid", "shield"].randomElement()!] = Int.random(in: 1...2)
            reward.diamonds = [8, 8, 12, 12, 16].randomElement()!
            if Int.random(in: 0..<100) < 12 { reward.character = ["nobita", "shizuka", "dorami"].randomElement()! }
        }
        if let character = reward.character, profile.unlockedCharacters.contains(character) {
            reward.character = nil
            reward.coins += kind == .silver ? 40 : (kind == .gold ? 80 : 120)
        }
        return reward
    }

    private func apply(_ reward: RewardBundle) {
        profile.coins += reward.coins
        profile.diamonds += reward.diamonds
        reward.items.forEach { profile.itemInventory[$0.key, default: 0] += $0.value }
        reward.chests.forEach { profile.chestInventory[$0.key.rawValue, default: 0] += $0.value }
        reward.keys.forEach { profile.keyInventory[$0.key.rawValue, default: 0] = min(99, profile.keyInventory[$0.key.rawValue, default: 0] + $0.value) }
        if let character = reward.character { profile.unlockedCharacters.insert(character) }
    }

    func settleRun(level: Int, success: Bool, stars: Int, score: Int, pickupCoins: Int, pickupDiamonds: Int, items: [String: Int], chests: [ChestKind: Int], usedLoadout: Set<String>) {
        var reward = RewardBundle(coins: pickupCoins, diamonds: pickupDiamonds, items: items, chests: chests)
        if !success && reward.coins == 0 && reward.diamonds == 0 && items.isEmpty && chests.isEmpty { reward.coins = 10 }
        if success {
            let base = level <= 5 ? 40 : (level <= 12 ? 70 : (level <= 20 ? 100 : 130))
            reward.coins += base + (stars == 2 ? 25 : (stars == 3 ? 55 : 0))
            let first = !profile.firstClears.contains(level)
            if first {
                reward.coins += level <= 5 ? 30 : (level <= 12 ? 50 : (level <= 20 ? 70 : 90))
                reward.diamonds += level <= 5 ? 3 : (level <= 12 ? 5 : (level <= 20 ? 7 : 9))
                profile.firstClears.insert(level)
            }
            if stars == 3 {
                let chance = level <= 5 ? 8 : (level <= 12 ? 12 : (level <= 20 ? 15 : 18))
                if Int.random(in: 0..<100) < chance { reward.diamonds += 1 }
            }
            profile.unlockedLevel = max(profile.unlockedLevel, min(20, level + 1))
            profile.levelStars[String(level)] = max(profile.levelStars[String(level)] ?? 0, stars)
            applyMilestones(into: &reward)
        }
        apply(reward)
        for id in profile.loadout where !usedLoadout.contains(id) { profile.itemInventory[id, default: 0] += 1 }
        profile.loadout.removeAll()
        let summary = RunSummary(level: level, success: success, stars: stars, score: score, coins: reward.coins, diamonds: reward.diamonds, itemRewards: reward.items, chestRewards: Dictionary(uniqueKeysWithValues: reward.chests.map { ($0.key.rawValue, $0.value) }), keyRewards: Dictionary(uniqueKeysWithValues: reward.keys.map { ($0.key.rawValue, $0.value) }))
        profile.lastRunSummary = summary
        latestResult = summary
        save()
        audio.playSFX(success ? "sfx_win" : "sfx_hurt", settings: profile.settings)
        navigate(.result)
    }

    private func applyMilestones(into reward: inout RewardBundle) {
        let clears = profile.firstClears.count
        let milestones: [(Int, ChestKind, KeyKind, Int)] = [(3,.wood,.copper,0),(5,.wood,.copper,10),(8,.silver,.silver,15),(12,.silver,.silver,20),(16,.gold,.gold,25),(20,.gold,.gold,30),(25,.purple,.purple,40)]
        for (count, chest, key, diamonds) in milestones where clears >= count && !profile.claimedMilestones.contains(count) {
            profile.claimedMilestones.insert(count)
            reward.chests[chest, default: 0] += 1
            reward.keys[key, default: 0] += 1
            reward.diamonds += diamonds
        }
    }

    func showToast(_ message: String) {
        toast = message
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in if self?.toast == message { self?.toast = nil } }
    }

    func save() { SaveStore.save(profile) }
}

File: 哆啦A梦-Android-cocos/Doraemon-Parkour-game/assets/scripts/core/GameAppState.ts
import { EventTarget } from 'cc';
import {
    ChestKind,
    RewardBundle,
    SaveProfile,
    chestByKind,
    characterById,
    itemById,
    keyByKind,
    ownedCharacterFallbackCoins,
    rollChest,
    settleEconomy,
} from '../data/GameData';
import { SaveStore } from './SaveStore';
import { haptic } from './Haptics';
import { RunOutcome } from '../run/RunWorld';

/** 主路由页面（对齐总结 §五 状态机；M0 只定义枚举，页面按里程碑接入） */
export enum Screen {
    Loading = 'loading',
    Home = 'home',
    Levels = 'levels',
    Run = 'run',
    Result = 'result',
    Inventory = 'inventory',
    Shop = 'shop',
    Settings = 'settings',
}

export interface ModalEntry {
    id: number;
    kind: ModalKind;
    payload?: unknown;
}

export interface ToastEntry {
    id: number;
    message: string;
}

/** 弹窗类型（M1：设置/关于/隐私/充值；后续里程碑再扩展） */
export enum ModalKind {
    Settings = 'settings',
    About = 'about',
    Privacy = 'privacy',
    Recharge = 'recharge',
    Pause = 'pause',
    /** 局内背包遮罩（对齐 iOS showInventoryOverlay） */
    Inventory = 'inventory',
    ChestReward = 'chest-reward',
    ChestFail = 'chest-fail',
    ShopCharacter = 'shop-character',
    ShopItem = 'shop-item',
    ShopChest = 'shop-chest',
    ShopKey = 'shop-key',
}

/** 结算结果（成败/星级/得分；奖励卡含数量） */
export interface RunResult {
    level: number;
    success: boolean;
    stars: number;
    score: number;
    /** 局内获得且未使用的道具/宝箱卡 */
    rewardCards?: Array<{ path: string; count: number }>;
}

/**
 * 全局状态单例（对齐规范 §9）：主路由、弹层栈、Toast、存档入口。
 * 事件：screen-changed / modal-changed / toast-changed / profile-changed。
 */
export class GameAppState {
    private static _instance: GameAppState | null = null;

    static get instance(): GameAppState {
        if (!this._instance) {
            this._instance = new GameAppState();
        }
        return this._instance;
    }

    readonly events = new EventTarget();
    readonly saveStore = new SaveStore();

    profile: SaveProfile;
    screen: Screen = Screen.Loading;
    /** 加载页用途：开机进首页 / 进局预加载 */
    loadingPurpose: 'boot' | 'run' = 'boot';
    selectedLevel = 1;
    runSessionID = '';
    latestResult: RunResult | null = null;

    private _modals: ModalEntry[] = [];
    private _toasts: ToastEntry[] = [];
    private _nextId = 1;

    constructor() {
        this.profile = this.saveStore.load();
    }

    setScreen(screen: Screen): void {
        if (this.screen === screen) {
            return;
        }
        this.screen = screen;
        this.events.emit('screen-changed', screen);
    }

    get modals(): readonly ModalEntry[] {
        return this._modals;
    }

    pushModal(kind: ModalKind, payload?: unknown): number {
        const id = this._nextId++;
        this._modals.push({ id, kind, payload });
        this.events.emit('modal-changed');
        return id;
    }

    popModal(): void {
        this._modals.pop();
        this.events.emit('modal-changed');
    }

    openSettings(): void {
        this.pushModal(ModalKind.Settings);
    }

    openAbout(): void {
        this.pushModal(ModalKind.About);
    }

    openPrivacy(): void {
        this.pushModal(ModalKind.Privacy);
    }

    openRecharge(): void {
        this.pushModal(ModalKind.Recharge);
    }

    openPause(): void {
        this.pushModal(ModalKind.Pause);
    }

    /**
     * 打开背包：局内 → 遮罩暂停；首页等 → 整页背包（对齐 iOS openInventory）
     */
    openInventory(from: 'home' | 'run' = 'home'): void {
        if (from === 'run') {
            this.pushModal(ModalKind.Inventory);
        } else {
            this.clearModals();
            this.setScreen(Screen.Inventory);
        }
    }

    /** 关闭局内背包遮罩 */
    closeInventoryOverlay(): void {
        if (this.topModalKind() === ModalKind.Inventory) {
            this.popModal();
            return;
        }
        // 若设置等叠在背包上，整栈清到去掉 Inventory
        const idx = this._modals.findIndex((m) => m.kind === ModalKind.Inventory);
        if (idx >= 0) {
            this._modals = this._modals.slice(0, idx);
            this.events.emit('modal-changed');
        }
    }

    clearModals(): void {
        this._modals = [];
        this.events.emit('modal-changed');
    }

    topModalKind(): ModalKind | null {
        const top = this._modals[this._modals.length - 1];
        return top ? top.kind : null;
    }

    /** 进关：直接进 Run（素材异步贴图；不再插加载页） */
    startLevel(level: number): void {
        this.selectedLevel = level;
        this.runSessionID = `run_${Date.now()}_${this._nextId++}`;
        this.screen = Screen.Run;
        this.events.emit('screen-changed', Screen.Run);
    }

    /**
     * 同关重开：强制重建 Run 页（即使当前已在 Run 也要 emit）。
     * 与 iOS `runSessionID = UUID(); navigate(.run)` 一致；关卡种子保证布局回到开局。
     */
    restartLevel(): void {
        this.clearModals();
        this.runSessionID = `run_${Date.now()}_${this._nextId++}`;
        this.screen = Screen.Run;
        this.events.emit('screen-changed', Screen.Run);
    }

    /** 结算入账（总结 §二/§2.8：拾取折算+通关/星级/首通/节点+道具退回；M6 收口） */
    settleRun(result: RunOutcome): void {
        const s = settleEconomy(result, this.profile);
        const p = this.profile;
        p.coins += s.coins;
        p.diamonds += s.diamonds;
        for (const id of s.itemsAdded) {
            p.itemInventory[id] = (p.itemInventory[id] ?? 0) + 1;
        }
        for (const kind of s.chestsAdded) {
            p.chestInventory[kind] = (p.chestInventory[kind] ?? 0) + 1;
        }
        for (const keyKind of s.keysAdded) {
            p.keyInventory[keyKind] = Math.min(99, (p.keyInventory[keyKind] ?? 0) + 1);
        }
        if (result.success) {
            p.unlockedLevel = Math.max(p.unlockedLevel, Math.min(20, result.level + 1));
            p.levelStars[`${result.level}`] = Math.max(p.levelStars[`${result.level}`] ?? 0, result.stars);
            p.firstClear[`${result.level}`] = true;
            for (const id of s.claimedNodes) {
                p.nodeClaims[id] = true;
            }
        }
        // 未使用的携带道具退回背包并清空携带（已消耗的不退）
        for (const id of p.loadout) {
            p.itemInventory[id] = (p.itemInventory[id] ?? 0) + 1;
        }
        p.loadout = [];
        p.lastRunSummary = {
            level: result.level,
            success: result.success,
            stars: result.stars,
            score: result.score,
            coins: s.coins,
            diamonds: s.diamonds,
        };
        this.commitProfile();
        this.latestResult = {
            level: result.level,
            success: result.success,
            stars: result.stars,
            score: result.score,
            rewardCards: s.rewardCards.slice(0, 9),
        };
        this.setScreen(Screen.Result);
    }

    /** 开箱（总结 §3.11.1 / §2.9）：先判箱量再判钥匙；成功扣 1+1 并掉落入账 */
    openChest(kind: ChestKind): void {
        const chest = chestByKind(kind);
        const boxCount = this.profile.chestInventory[kind] ?? 0;
        const vibe = this.profile.settings.vibrationEnabled;
        if (boxCount <= 0) {
            haptic('warning', vibe);
            this.pushModal(ModalKind.ChestFail, { reason: 'missingChest', kind });
            return;
        }
        const keyCount = this.profile.keyInventory[chest.key] ?? 0;
        if (keyCount <= 0) {
            haptic('warning', vibe);
            this.pushModal(ModalKind.ChestFail, { reason: 'missingKey', kind });
            return;
        }
        this.profile.chestInventory[kind] = boxCount - 1;
        this.profile.keyInventory[chest.key] = keyCount - 1;

        let reward = rollChest(kind, this.profile.unlockedCharacters);
        if (reward.character && this.profile.unlockedCharacters.includes(reward.character)) {
            reward = { ...reward, coins: reward.coins + ownedCharacterFallbackCoins(kind), character: null };
        }
        this.applyReward(reward);
        this.commitProfile();
        // 成功震动在 ChestRewardModal 播放（对齐 iOS notification.success）
        this.pushModal(ModalKind.ChestReward, { kind, reward });
    }

    private applyReward(reward: RewardBundle): void {
        this.profile.coins += reward.coins;
        this.profile.diamonds += reward.diamonds;
        for (const id of reward.items) {
            this.profile.itemInventory[id] = (this.profile.itemInventory[id] ?? 0) + 1;
        }
        if (reward.character && !this.profile.unlockedCharacters.includes(reward.character)) {
            this.profile.unlockedCharacters.push(reward.character);
        }
    }

    /** 商城购买入口（总结 §5.4：先详情再扣费；一件商品一种货币） */
    buyCharacter(id: string): void {
        const c = characterById(id);
        const vibe = this.profile.settings.vibrationEnabled;
        if (this.profile.unlockedCharacters.includes(id)) {
            this.toast('已拥有该角色');
            return;
        }
        if (!c.availableInRun) {
            this.toast('动作素材待补…');
            return;
        }
        if (this.profile.diamonds < c.price) {
            haptic('warning', vibe);
            this.toast('钻石不足');
            return;
        }
        this.profile.diamonds -= c.price;
        this.profile.unlockedCharacters.push(id);
        this.commitProfile();
        haptic('success', vibe);
        this.toast(`购买成功，已解锁 ${c.name}`);
        this.popModal();
    }

    buyItem(id: string): void {
        const item = itemById(id);
        const vibe = this.profile.settings.vibrationEnabled;
        if (this.profile.coins < item.buyPrice) {
            haptic('warning', vibe);
            this.toast('金币不足');
            return;
        }
        this.profile.coins -= item.buyPrice;
        this.profile.itemInventory[id] = (this.profile.itemInventory[id] ?? 0) + 1;
        this.commitProfile();
        haptic('success', vibe);
        this.toast(`购买成功 +1 ${item.name}`);
        this.popModal();
    }

    buyChest(kind: ChestKind): void {
        const chest = chestByKind(kind);
        const vibe = this.profile.settings.vibrationEnabled;
        const wallet = chest.currency === 'coins' ? this.profile.coins : this.profile.diamonds;
        if (wallet < chest.price) {
            haptic('warning', vibe);
            this.toast(chest.currency === 'coins' ? '金币不足' : '钻石不足');
            return;
        }
        if (chest.currency === 'coins') {
            this.profile.coins -= chest.price;
        } else {
            this.profile.diamonds -= chest.price;
        }
        this.profile.chestInventory[kind] = (this.profile.chestInventory[kind] ?? 0) + 1;
        this.commitProfile();
        haptic('success', vibe);
        this.toast(`购买成功 +1 ${chest.name}`);
        this.popModal();
    }

    buyKey(kind: ChestKind): void {
        const chest = chestByKind(kind);
        const key = keyByKind(chest.key);
        const vibe = this.profile.settings.vibrationEnabled;
        const wallet = key.currency === 'coins' ? this.profile.coins : this.profile.diamonds;
        const owned = this.profile.keyInventory[chest.key] ?? 0;
        if (owned >= 99) {
            this.toast('钥匙已达上限 99');
            return;
        }
        if (wallet < key.price) {
            haptic('warning', vibe);
            this.toast(key.currency === 'coins' ? '金币不足' : '钻石不足');
            return;
        }
        if (key.currency === 'coins') {
            this.profile.coins -= key.price;
        } else {
            this.profile.diamonds -= key.price;
        }
        this.profile.keyInventory[chest.key] = owned + 1;
        this.commitProfile();
        haptic('success', vibe);
        this.toast(`购买成功 +1 ${key.name}`);
        this.popModal();
    }

    get toasts(): readonly ToastEntry[] {
        return this._toasts;
    }

    toast(message: string): void {
        const id = this._nextId++;
        this._toasts.push({ id, message });
        this.events.emit('toast-changed');
        setTimeout(() => {
            const idx = this._toasts.findIndex((t) => t.id === id);
            if (idx >= 0) {
                this._toasts.splice(idx, 1);
                this.events.emit('toast-changed');
            }
        }, 2000);
    }

    /** 存档成功后统一通知 UI 刷新 */
    commitProfile(): void {
        this.saveStore.save(this.profile);
        this.events.emit('profile-changed');
    }

    /** 清缓存（清结算摘要、保留进度） */
    clearCache(): void {
        this.saveStore.clearCache(this.profile);
        this.events.emit('profile-changed');
    }

    /** 重置为新档 */
    resetProfile(): void {
        this.profile = this.saveStore.reset();
        this.events.emit('profile-changed');
    }
}
